[SIL] Hollow out Builtin.copy and deprecate _copy.

nate-chandler opened the pull request:
The copy operator has been implemented and doesn't use it.  Remove `Builtin.copy` and `_copy` as much as currently possible.

Source compatibility requires that `_copy` remain in the stdlib.  It is deprecated here and just uses the copy operator.

Handling old swiftinterfaces requires that `Builtin.copy` be defined.  Redefine it here as a passthrough--SILGen machinery will produce the necessary copy_addr.

rdar://127502242

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

atrick reviewed (approved):
Seems like we should be able to remove this.

nate-chandler commented:
@lorentey @lorentey Is it safe to remove the `_copy` function?

tshortli commented:
It looks like the main concern is source compatibility with the removal of the always-emit-into-client `_copy` decl; I don't see any other changes that would affect clients of the stdlib. Hopefully there aren't any projects we can't break that have adopted it, but I can't evaluate that with confidence.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

lorentey commented on stdlib/public/core/LifetimeManager.swift:
```diff
@@ -265,3 +265,12 @@ extension String {
 }
 #endif
+
+@available(swift, deprecated: 6, message: "Use the copy operator")
```
It would be better to rather deprecate this in all language modes, not just in Swift 6+:

```suggestion
@available(*, deprecated, message: "Use the copy operator")
```

nate-chandler replied:
Fixed

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
```
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Swift.swiftmodule/x86_64-apple-macos.swiftinterface:12443:5: error: module 'Builtin' has no member named 'copy'
12441 | @_alwaysEmitIntoClient @inlinable @_transparent @_semantics("lifetimemanagement.copy") public func _copy<T>(_ value: T) -> T {
12442 |   #if $BuiltinCopy
12443 |     Builtin.copy(value)
      |     `- error: module 'Builtin' has no member named 'copy'
12444 |   #else
12445 |     value
```

nate-chandler committed:
[SIL] Hollow out Builtin.copy, deprecate _copy.

The copy operator has been implemented and doesn't use it.  Remove
`Builtin.copy` and `_copy` as much as currently possible.

Source compatibility requires that `_copy` remain in the stdlib.  It is
deprecated here and just uses the copy operator.

Handling old swiftinterfaces requires that `Builtin.copy` be defined.
Redefine it here as a passthrough--SILGen machinery will produce the
necessary copy_addr.

rdar://127502242

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please apple silicon benchmark

nate-chandler commented:
Source compat failures match failures in baseline.
